storage_badger.go

```diff
@@ -0,0 +1,160 @@
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+
+*/
+
+package verifiabledatastructures
+
+import (
+	"encoding/hex"
+	"path/filepath"
+	"sync"
+
+	"golang.org/x/net/context"
+
+	"github.com/dgraph-io/badger"
+	"github.com/golang/protobuf/proto"
+)
+
+// BadgerBackedService gives a service that persists to a BoltDB file.
+type BadgerBackedService struct {
+	// Path is a path to a directory where we will create files, one per user Map / user Log.
+	Path string
+
+	dbLock sync.RWMutex
+	dbs    map[string]*badger.DB
+}
+
+// Close calls the underlying Close method on the BoltDBs which releases file locks.
+func (bbs *BadgerBackedService) Close() {
+	bbs.dbLock.Lock()
+	defer bbs.dbLock.Unlock()
+
+	for _, db := range bbs.dbs {
+		// Ignore any errors
+		db.Close()
+	}
+}
+
+// Returns nil if not found
+func (bbs *BadgerBackedService) getDB(key string) *badger.DB {
+	bbs.dbLock.RLock()
+	defer bbs.dbLock.RUnlock()
+
+	if bbs.dbs == nil {
+		return nil
+	}
+
+	rv, ok := bbs.dbs[key]
+	if ok {
+		return rv
+	}
+
+	return nil
+}
+
+// getOrCreateDB gets and/or creates DB
+func (bbs *BadgerBackedService) getOrCreateDB(ns []byte) (*badger.DB, error) {
+	key := hex.EncodeToString(ns)
+
+	// First try with simple read lock
+	rv := bbs.getDB(key)
+	if rv != nil {
+		return rv, nil
+	}
+
+	// Else, get a write lock and load it up
+	bbs.dbLock.Lock()
+	defer bbs.dbLock.Unlock()
+
+	if bbs.dbs == nil {
+		bbs.dbs = make(map[string]*badger.DB)
+	}
+
+	// Check again, as may have been added since last
+	rv, ok := bbs.dbs[key]
+	if ok {
+		return rv, nil
+	}
+
+	// Get or create on distk
+	opts := badger.DefaultOptions
+	dir := filepath.Join(bbs.Path, key)
+	opts.Dir = dir
+	opts.ValueDir = dir
+	db, err := badger.Open(opts)
+	if err != nil {
+		return nil, err
+	}
+	bbs.dbs[key] = db
+
+	return db, nil
+}
+
+// ExecuteReadOnly executes a read only query
+func (bbs *BadgerBackedService) ExecuteReadOnly(ctx context.Context, namespace []byte, f func(ctx context.Context, db KeyReader) error) error {
+	db, err := bbs.getOrCreateDB(namespace)
+	if err != nil {
+		return err
+	}
+	return db.View(func(tx *badger.Txn) error {
+		return f(ctx, &badgerReaderWriter{Tx: tx})
+	})
+}
+
+// ExecuteUpdate executes an update query
+func (bbs *BadgerBackedService) ExecuteUpdate(ctx context.Context, namespace []byte, f func(ctx context.Context, db KeyWriter) error) error {
+	db, err := bbs.getOrCreateDB(namespace)
+	if err != nil {
+		return err
+	}
+	return db.Update(func(tx *badger.Txn) error {
+		return f(ctx, &badgerReaderWriter{Tx: tx})
+	})
+}
+
+type badgerReaderWriter struct {
+	Tx *badger.Txn
+}
+
+func (db *badgerReaderWriter) Get(ctx context.Context, bucket, key []byte, value proto.Message) error {
+	var bucketKey []byte
+	bucketKey = append(bucketKey, bucket...)
+	bucketKey = append(bucketKey, key...)
```

Comment:
At first glance this bit worries me a bit as bucket `foo` and key `bar` would store to the same location as bucket `foob` key `ar` etc.

In practice, the bucket names are currently ASCII text and there's only 15 or so different ones used, however the keys are pretty arbitrary and I think it would be safer to not assume that we would never produce a combination like that.

Maybe something like the following would guarantee that no 2 combinations would produce the same `bucketKey`:

```golang
bucketKey := append([]byte(url.PathEscape(string(bucket))+"/"), key...)
```

Reply:
I see the potential issue here, however, I glanced at the source for `url.PathEscape` and I looks kind of heavy for a section of code that gets called on every read/write.  Indeed i'm not sure it even helps in this case, since the escaping still results in a fixed set of 15 unique bucket names.

Thinking about whether a lighter weight solution could achieve the same: it should be sufficient to use a single character separator (like "/"). As long as the separator never appears in any of the bucket names, it is guaranteed to be the first instance of that byte. Even if "/" is the first byte in some of the keys, I don't think it would ever be possible to have two bucketKey's conflict.

This would also allow us to avoid the conversion back and forth from []byte->string->[]byte.

I'll push up a proposal for your review, but if you don't agree, or there is another reason I'm not seeing that the escaping is required, I'll go your way instead.

Reply:
@jvsteiner - why don't I change the interface to just not deal with buckets? :)

Give me a few mins and I'll shoot through a PR that changes `KeyReader` / `KeyWriter` to just not need to deal with it at all - I think that'll make it a bit easier.

Reply:
even better!